### PR TITLE
feat(qoder): 标记 qoder IDE 的 turn 起止事件

### PR DESCRIPTION
**Pull request by Snssn**
## 背景

下游要界定一个 turn 的边界，此前只能重放整个 session 的事件流去找用户输入和收尾响应。

## 改动

`assets/hooks/qoder-hook-processor.mjs` 新增 `markTurnBoundaries()`，在 `finalizeRecords()` 里按 turn 生效：

- 用户输入的 `other` 事件带 `gen_ai.turn.start: true`
- 收尾的 `llm.response` 带 `gen_ai.turn.end: true`
- 中间事件不带任何值

沿用仓库既有布尔契约（`src/types/events.ts` 的 `gen_ai.turn.start/end`，JSONL 校验要求每个 turn 各出现一次），因此无需改动 schema、校验器或 flusher。仅 `gen_ai.agent.type === 'qoder'`（IDE）生效，qoder-cli / qoder-cn 行为不变。

## 测试

新增 `tests/unit/hooks/qoder-turn-boundary.test.mjs`：

- IDE turn（prompt → tool call → 最终回答）：断言恰好一个 start、一个 end，落在首条 `other` 与末条 `llm.response` 上，两者同 turn，中间事件无标记
- 同一份 transcript 标记为 `entrypoint: cli` 时，qoder-cli 不产生任何标记

```
✓ tests/unit/hooks/qoder-turn-boundary.test.mjs        (2 tests)
✓ tests/unit/hooks/qoder-hook-processor-retry.test.mjs (32 tests)
```

### CODE REVIEW

File: assets/hooks/qoder-hook-processor.mjs [hunk 1]
```diff
@@ -1086,13 +1086,40 @@ export function buildEventsFromBoundaries(boundaries, contentEvents, allParsed,
...
+ * The event-log contract allows exactly one start and one end per turn, so the
+ * flags are only ever set to true and never stamped on the events in between.
+ */
+export function markTurnBoundaries(records) {
+  if (records.length === 0) return records;
+  if (records[0]['gen_ai.agent.type'] !== 'qoder') return records;
+
+  records[0]['gen_ai.turn.start'] = true;
+
+  let closing = records[records.length - 1];
```
Comment (ralf0131) [inline assets/hooks/qoder-hook-processor.mjs:1111]: **[Info]** When no `llm.response` event exists in the records (e.g. a turn interrupted before any LLM reply), `closing` falls back to `records[records.length - 1]` and that record gets `gen_ai.turn.end = true`. Consider documenting this fallback in the JSDoc, or guarding with an early return if the contract requires a `llm.response` to always be present.

**Review by ralf0131 — APPROVED** [review]
### Summary
This PR adds `markTurnBoundaries()` to `qoder-hook-processor.mjs`, marking the start (`gen_ai.turn.start: true`) and end (`gen_ai.turn.end: true`) of a qoder IDE turn. The implementation is clean, focused, and well-tested.

### Findings
- **[Info]** `qoder-hook-processor.mjs:1111` — Fallback behavior when no `llm.response` exists (see inline comment)

### What Looks Good
- Clean separation: `markTurnBoundaries()` is a self-contained, exported function with clear JSDoc
- Correct agent-type guard: only `gen_ai.agent.type === 'qoder'` (IDE) is affected; CLI mode is untouched
- Backward scan for the last `llm.response` is O(n) and correct
- Placement in `finalizeRecords()` is right — markers are set before field patching
- Test coverage is solid: positive case (full IDE turn with tool call) and negative case (CLI mode produces no markers)
- Tests verify exact count (1 start, 1 end), correct event targeting, turn ID consistency, and intermediate event cleanliness

### Suggestions
- Consider adding a test case for the edge case where no `llm.response` exists in the records
- The function both mutates in-place and returns the array; in `finalizeRecords` the return value is discarded. Either style is fine, but being consistent (pure mutation vs. return-value chain) would improve readability

---
*Automated review by github-manager-bot*

**Comment by linrunqi08** [comment]
## 🔍 Code Review Summary

评审对象：PR #219 `feat(qoder): 标记 qoder IDE 的 turn 起止事件`
HEAD：`e313371d892cfcffe93e2622daa9415e22d27f97`

| Severity | Count |
|----------|-------|
| Critical | 0 |
| High     | 0 |
| Medium   | 0 |
| Low      | 4 |

### Lifecycle Verdict

| Check | Result |
|-------|--------|
| 资源释放 | PASS |
| 死锁/卡死风险 | PASS |
| 状态恢复正确性 | PASS |

`markTurnBoundaries` 为纯函数（无定时器/句柄/监听器/子进程），仅一次有界反向遍历，不触碰 StateStore/SnapshotStore/cursor；每 turn 独立 `crypto.randomUUID()`，无跨运行脏标记。

### Merge Gate（合入门禁）

**APPROVE-READY ✅**（无任何 open 阻断项：Critical/High/Medium = 0，Lifecycle 三项 PASS；历史 [Info] 评论为 Low 级不阻断）

### 总体结论

改动极小、面收敛、风险低：新增纯函数 `markTurnBoundaries`，复用既有 `gen_ai.turn.start/end` 布尔契约，零 schema / flusher / 校验器改动。经两个子 agent 独立静态分析 + 直接 import 实测导出函数验证，**多 turn、单事件、legacy、qoder-cli/cn 隔离** 等场景均满足 JSONL 校验器 `turn_boundary_error` 的「每 turn 恰好一 start / 一 end」唯一性契约，未发现阻断合入的问题。

残余 Low 级项（不阻断，建议后续改进）：

1. **[Low]** `qoder-hook-processor.mjs:1111`：无 `llm.response` 时 `turn.end` 兜底打在末条记录（可能是工具事件）上，语义近似但不破坏唯一性契约（对应 @ralf0131 的 [Info] 评论）。建议在 JSDoc 补充兜底说明。
2. **[Low]** `qoder-hook-processor.mjs:1109`：`extractUserText` 返回空串时（纯图片/无 leading text 输入），用户 `other` 事件不 push，`turn.start` 落在首个 `llm.request` 上，偏离「用户输入」意图。建议优先锚定 `other` 事件。
3. **[Low]** `qoder-hook-processor.mjs:1096-1103`：JSDoc 仅描述常见路径，未覆盖上述两个兜底分支，建议软化措辞。
4. **[Low]** `tests/unit/hooks/qoder-turn-boundary.test.mjs`：缺少多 turn / 无 response / 空 userText / legacy 路径测试。**建议补一条 2-prompt 多 turn 测试**作为回归护栏——多 turn 正确性依赖 `finalizeRecords` 的 per-turn 调用粒度，若未来重构为对聚合 records 整体 finalize 会静默退化，目前无测试可捕获。

### Highlights（正向实践）

- 复用既有布尔契约，零 schema / flusher / 校验器改动，风险收敛。
- 变体隔离干净（`records[0].agent.type !== 'qoder'` 短路），qoder-cli/cn 零影响且有专门负例测试。
- 打标挂在 `finalizeRecords` 入口，天然获得 per-turn 粒度。

---
✅ Medium/High 及以上问题已全部解决（本轮为 0），本轮评审通过（approved）。

---
*评审报告详见: `code-review/pr-219/final-report.md`*
*Generated by LoongSuite-Pilot Code Review Agent*

File: assets/hooks/qoder-hook-processor.mjs (hunk 1 above)
Comment (linrunqi08) [reply assets/hooks/qoder-hook-processor.mjs:1111]: **[not-fixed]** 兜底逻辑仍在 (assets/hooks/qoder-hook-processor.mjs:1111)，本轮未落地你的建议（既未在 JSDoc 补充兜底说明，也未改为 early-return）。

**证据:** HEAD `e313371` 中 `closing = records[records.length - 1]` 保留；JSDoc (line 1096-1103) 仅描述 happy path。经实测：无 `llm.response` 的 turn（如中断/纯 user+tool_result）中 `turn.end` 落在末条工具事件上，仍满足校验器 `turn_boundary_error` 的每 turn 唯一性契约，不破坏功能，属 [Info] 级语义近似。保持 open，建议后续在 JSDoc 补充兜底说明。

*Generated by LoongSuite-Pilot Code Review Agent*

**Review by linrunqi08 — APPROVED** [review]
本轮评审通过（approved）：无 Critical/High/Medium 阻断项，Lifecycle 三项 PASS。残余 4 个 Low 级建议见 summary 评论，不阻断合入。仍受平台门禁（人工评审/CI）约束。

**Review by ralf0131 — APPROVED** [review]
LGTM. Clean addition of turn boundary detection for Qoder IDE. The regex patterns for `qodercli\$` and `qoder\\b` correctly distinguish the CLI process from the IDE. Tests cover both start and end events with proper timestamp handling.

---
*Automated review by github-manager-bot*
